src/ucp/core/ucp_listener.c

```diff
@@ -205,75 +228,118 @@ ucs_status_t ucp_listener_create(ucp_worker_h worker,
+        listener->accept_cb = params->accept_handler.cb;
+        listener->arg       = params->accept_handler.arg;
+    } else if (params->field_mask & UCP_LISTENER_PARAM_FIELD_CONN_HANDLER) {
+        UCP_CHECK_PARAM_NON_NULL(params->conn_handler.cb, status,
+                                 goto err_free_listener);
+        listener->conn_cb   = params->conn_handler.cb;
+        listener->arg       = params->conn_handler.arg;
+    }
 
-        iface_params.field_mask                     = UCT_IFACE_PARAM_FIELD_OPEN_MODE |
-                                                      UCT_IFACE_PARAM_FIELD_SOCKADDR;
-        iface_params.open_mode                      = UCT_IFACE_OPEN_MODE_SOCKADDR_SERVER;
-        iface_params.mode.sockaddr.conn_request_cb  = ucp_listener_conn_request_callback;
-        iface_params.mode.sockaddr.conn_request_arg = listener;
-        iface_params.mode.sockaddr.listen_sockaddr  = params->sockaddr;
-        iface_params.mode.sockaddr.cb_flags         = UCT_CB_FLAG_ASYNC;
-
-        status = ucp_worker_iface_open(worker, tl_id, &iface_params,
-                                       &listener->wiface);
-        if (status != UCS_OK) {
-            goto err_free;
-        }
+    status = ucs_sockaddr_get_port(params->sockaddr.addr, &port);
+    if (status != UCS_OK) {
+       goto err_free_listener;
+    }
 
-        status = ucp_worker_iface_init(worker, tl_id, &listener->wiface);
-        if ((status != UCS_OK) ||
-            ((context->config.features & UCP_FEATURE_WAKEUP) &&
-            !(listener->wiface.attr.cap.flags & UCT_IFACE_FLAG_CB_ASYNC))) {
-            ucp_worker_iface_cleanup(&listener->wiface);
-            ucs_free(listener);
-            continue;
+    /* Go through all the available resources and for each one, check if the given
+     * sockaddr is accessible from its md. Start listening on all the mds that
+     * satisfy this.
+     * If the given port is set to 0, i.e. use a random port, the first transport
+     * in the sockaddr priority list from the environment configuration will
+     * dictate the port to listen on for the other sockaddr transports in the list.
+     * */
+    for (i = 0; i < context->config.num_sockaddr_tls; i++) {
+        ucs_for_each_bit(tl_id, context->tl_bitmap) {
+            resource = &context->tl_rscs[tl_id];
+            tl_md    = &context->tl_mds[resource->md_index];
+
+            if (strcmp(context->config.sockadrr_tls[i].cm_tl_name,
+                       resource->tl_rsc.tl_name) ||
+                !(tl_md->attr.cap.flags & UCT_MD_FLAG_SOCKADDR) ||
+                !uct_md_is_sockaddr_accessible(tl_md->md, &params->sockaddr,
+                                               UCT_SOCKADDR_ACC_LOCAL)) {
+                continue;
+            }
+
+            tmp = ucs_realloc(listener->wifaces,
+                              sizeof(*listener->wifaces) * (sockaddr_tls + 1),
+                              "listener wifaces");
+            if (tmp == NULL) {
```

Comment:
need to free old memory to handle `realloc` error correctly

Reply:
@evgeny-leksikov but the old memery is freed in the err_free_listener_wifaces label.